pkg/micro/ocdav/service.go

```diff
@@ -0,0 +1,173 @@
+	"go-micro.dev/v4"
+	"go-micro.dev/v4/registry"
+	"go-micro.dev/v4/server"
+	"go.opentelemetry.io/otel/propagation"
+)
+
+const (
+	// ServerName to use when announcing the service to the registry
+	ServerName = "ocdav"
+)
+
+// Service initializes the ocdav service and underlying http server.
+func Service(opts ...Option) (micro.Service, error) {
+
+	sopts := newOptions(opts...)
+
+	// set defaults
+	if err := setDefaults(&sopts); err != nil {
+		return nil, err
+	}
+
+	sopts.Logger = sopts.Logger.With().Str("name", sopts.Name).Logger()
+
+	srv := httpServer.NewServer(
+		server.TLSConfig(sopts.TLSConfig),
+		server.Name(sopts.Name),
+		server.Address(sopts.Address), // Address defaults to ":0" and will pick any free port
+	)
+
+	revaService, err := ocdav.NewWith(&sopts.config, sopts.FavoriteManager, sopts.lockSystem, &sopts.Logger)
+	if err != nil {
+		return nil, err
+	}
+
+	// register additional webdav verbs
+	chi.RegisterMethod(ocdav.MethodPropfind)
+	chi.RegisterMethod(ocdav.MethodProppatch)
+	chi.RegisterMethod(ocdav.MethodLock)
+	chi.RegisterMethod(ocdav.MethodUnlock)
+	chi.RegisterMethod(ocdav.MethodCopy)
+	chi.RegisterMethod(ocdav.MethodMove)
+	chi.RegisterMethod(ocdav.MethodMkcol)
+	chi.RegisterMethod(ocdav.MethodReport)
+	r := chi.NewRouter()
+
+	if err := useMiddlewares(r, &sopts, revaService); err != nil {
+		return nil, err
+	}
+
+	r.Handle("/*", revaService.Handler())
+
+	hd := srv.NewHandler(r)
+	if err := srv.Handle(hd); err != nil {
+		return nil, err
+	}
+
+	service := micro.NewService(
+		micro.Server(srv),
+		micro.Registry(registry.NewRegistry()),
+	)
+
+	// Init the service? make that optional?
+	service.Init()
+
+	// finally, return the service so it can be Run() by the caller himself
+	return service, nil
+}
+
+func setDefaults(sopts *Options) error {
+
+	// set defaults
+	if sopts.Name == "" {
+		sopts.Name = ServerName
+	}
+	if sopts.lockSystem == nil {
+		client, err := pool.GetGatewayServiceClient(sopts.config.GatewaySvc)
+		if err != nil {
+			return err
+		}
+		sopts.lockSystem = ocdav.NewCS3LS(client)
+	}
+	if sopts.FavoriteManager == nil {
+		sopts.FavoriteManager, _ = memory.New(map[string]interface{}{})
+	}
+	if !strings.HasPrefix(sopts.config.Prefix, "/") {
+		sopts.config.Prefix = "/" + sopts.config.Prefix
+	}
+	return nil
+}
+
+func useMiddlewares(r *chi.Mux, sopts *Options, svc global.Service) error {
+
+	// auth
+	for _, v := range svc.Unprotected() {
+		sopts.Logger.Info().Str("url", v).Msg("unprotected URL")
+	}
+	authMiddle, err := auth.New(map[string]interface{}{
+		"gatewaysvc": sopts.config.GatewaySvc,
+		"token_managers": map[string]interface{}{
+			"jwt": map[string]interface{}{
+				"secret": sopts.JWTSecret,
+			},
+		},
+	}, svc.Unprotected())
+	if err != nil {
+		return err
+	}
+
+	// log
+	lm := revaLogMiddleware.New()
+
+	// tracing
+	tm := func(h http.Handler) http.Handler { return h }
+	if sopts.TracingEnabled {
+		tm = traceHandler("ocdav", sopts.TracingCollector, sopts.TracingEndpoint)
```

Comment:
While this differs from the way rhttp (it wraps all middlewares), I think adding one trace middleware is the right approach. If other middlewares should be traced they should add their own span.